File: Pinecone/src/Pinecone/Core/Input.cpp
#include "pcpch.h"
#include "Input.h"

#include "Pinecone/Core/Application.h"

#include <GLFW/glfw3.h>

namespace Pinecone
{
	bool Input::IsKeyPressed(const KeyCode key)
	{
		// Get the GLFW window handle
		auto* window = static_cast<GLFWwindow*>(Application::Get().GetWindow().GetNativeWindow());
		// Get the state of the key and return if it is pressed
		auto state = glfwGetKey(window, static_cast<int32_t>(key));
		return state == GLFW_PRESS;
	}

	bool Input::IsMouseButtonPressed(const MouseCode button)
	{
		// Get the GLFW window handle
		auto* window = static_cast<GLFWwindow*>(Application::Get().GetWindow().GetNativeWindow());
		// Get the state of the mouse button and return if it is pressed
		auto state = glfwGetMouseButton(window, static_cast<int32_t>(button));
		return state == GLFW_PRESS;
	}

	bool Input::IsMouseButtonReleased(MouseCode button)
	{
		// Get the GLFW window handle
		auto* window = static_cast<GLFWwindow*>(Application::Get().GetWindow().GetNativeWindow());
		// Get the state of the mouse button and return if it is released
		auto state = glfwGetMouseButton(window, static_cast<int32_t>(button));
		return state == GLFW_RELEASE;
	}

	glm::vec2 Input::GetMousePosition()
	{
		// Get the GLFW window handle
		auto* window = static_cast<GLFWwindow*>(Application::Get().GetWindow().GetNativeWindow());

		// Get the position of the mouse cursors x and y
		double xpos, ypos;
		glfwGetCursorPos(window, &xpos, &ypos);

		// Return the x and y position as a 2D vector
		return { (float)xpos, (float)ypos };
	}

	float Input::GetMouseX()
	{
		return GetMousePosition().x;
	}

	float Input::GetMouseY()
	{
		return GetMousePosition().y;
	}
}

File: Pinecone-ScriptCore/Source/Pinecone/Input/Input.cs
namespace Pinecone
{
    public class Input
    {
        public static bool IsKeyDown(KeyCode keycode)
        {
            return InternalCalls.Input_IsKeyDown(keycode);
        }

        public static bool IsMouseButtonPressed(MouseCode button)
        {
            return InternalCalls.Input_IsMouseButtonPressed(button);
        }

        public static Vector2 GetMousePosition()
        {
            InternalCalls.Input_GetMousePosition(out Vector2 position);
            return position;
        }

        public static float GetMouseX()
        {
            InternalCalls.Input_GetMouseX(out float x);
            return x;
        }

        public static float GetMouseY()
        {
            InternalCalls.Input_GetMouseY(out float y);
            return y;
        }
    }
}
